# Dataset 3: Sales Data

We read in the csv file obtained from https://assets.datacamp.com/production/course_1294/datasets/sales.csvand use head() to peek at the data.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

sales = pd.read_csv('sales.csv')
sales.head()

Unnamed: 0              event_id        primary_act_id  \
0           1  abcaf1adb99a935fc661  43f0436b905bfa7c2eec   
1           2  6c56d7f08c95f2aa453c  1a3e9aecd0617706a794   
2           3  c7ab4524a121f9d687d2  4b677c3f5bec71eec8d1   
3           4  394cb493f893be9b9ed1  b1ccea01ad6ef8522796   
4           5  55b5f67e618557929f48  91c03a34b562436efa3c   

       secondary_act_id   purch_party_lkup_id  \
0  b85143bf51323b72e53c  7dfa56dd7d5956b17587   
1  f53529c5679ea6ca5a48  4f9e6fc637eaf7b736c2   
2  b85143bf51323b72e53c  6c2545703bd527a7144d   
3  b85143bf51323b72e53c  527d6b1eaffc69ddd882   
4  b85143bf51323b72e53c  8bd62c394a35213bdf52   

                                          event_name  \
0  Xfinity Center Mansfield Premier Parking: Flor...   
1      Gorge Camping - dave matthews band - sept 3-7   
2        Dodge Theatre Adams Street Parking - benise   
3  Gexa Energy Pavilion Vip Parking : kid rock wi...   
4                      Premier Parking - motley crue   

                           primary_act_name  secondary_act_name  \
0  XFINITY Center Mansfield Premier Parking                 NaN   
1                             Gorge Camping  Dave Matthews Band   
2                             Parking Event                 NaN   
3          Gexa Energy Pavilion VIP Parking                 NaN   
4  White River Amphitheatre Premier Parking                 NaN   

  major_cat_name minor_cat_name     ...     edu_1st_indv_val edu_2nd_indv_val  \
0           MISC        PARKING     ...                  NaN              NaN   
1           MISC        CAMPING     ...                  NaN              NaN   
2           MISC        PARKING     ...                  NaN              NaN   
3           MISC        PARKING     ...                  NaN              NaN   
4           MISC        PARKING     ...                  NaN              NaN   

  adults_in_hh_num  married_ind  child_present_ind home_owner_ind occpn_val  \
0              NaN          NaN                NaN            NaN       NaN   
1              NaN          NaN                NaN            NaN       NaN   
2              NaN          NaN                NaN            NaN       NaN   
3              NaN          NaN                NaN            NaN       NaN   
4              NaN          NaN                NaN            NaN       NaN   

  occpn_1st_val occpn_2nd_val dist_to_ven  
0           NaN           NaN         NaN  
1           NaN           NaN        59.0  
2           NaN           NaN         NaN  
3           NaN           NaN         NaN  
4           NaN           NaN         NaN  

[5 rows x 46 columns]

Like with the MOMA dataset, `head()` is not comprehensive enough to get a good glance at this wide dataset, so we follow up with `count()` to get further detail.

In [2]:
sales.count()

Unnamed: 0                5000
event_id                  5000
primary_act_id            5000
secondary_act_id          5000
purch_party_lkup_id       5000
event_name                5000
primary_act_name          5000
secondary_act_name        1586
major_cat_name            5000
minor_cat_name            5000
la_event_type_cat         5000
event_disp_name           5000
ticket_text               5000
tickets_purchased_qty     5000
trans_face_val_amt        5000
delivery_type_cd          5000
event_date_time           5000
event_dt                  5000
presale_dt                2108
onsale_dt                 4899
sales_ord_create_dttm     4996
sales_ord_tran_dt         5000
print_dt                  4576
timezn_nm                 5000
venue_city                5000
venue_state               5000
venue_postal_cd_sgmt_1    5000
sales_platform_cd         2578
print_flg                 5000
la_valid_tkt_event_flg    5000
fin_mkt_nm                5000
web_session_cookie_val    5000
gndr_cd 

We hone in on "presale_dt" and "onsale_dt" columns, to attempt to determine the average number of days in advance that presale tickets could be ordered.  We create a new dataframe with these columns.

In [3]:
presale = sales[['presale_dt', 'onsale_dt']]
presale

presale_dt   onsale_dt
0            NaN  2015-05-15
1            NaN  2009-03-13
2            NaN  2006-02-25
3            NaN  2011-04-22
4     2005-03-02  2005-03-04
5            NaN  2012-04-11
6            NaN  2006-05-12
7            NaN  2013-02-16
8            NaN  2012-08-18
9            NaN  2014-12-12
10           NaN  2006-06-12
11    2015-12-09  2015-12-11
12           NaN  2013-01-10
13           NaN  2006-11-29
14           NaN  2006-09-16
15    2008-04-03  2008-04-05
16    2015-03-19  2015-03-20
17    2008-05-14  2008-05-17
18           NaN  2006-08-28
19           NaN  2010-05-22
20           NaN  2015-02-20
21           NaN  2013-04-01
22           NaN         NaN
23           NaN  2008-06-07
24           NaN  2008-03-01
25           NaN         NaN
26    2015-02-04  2015-02-06
27           NaN  2012-03-01
28           NaN  2008-06-14
29    2016-01-12  2016-01-16
...          ...         ...
4970         NaN  2008-02-29
4971         NaN  2013-04-09
4972         NaN  2009-11-07
4973  2008-05-14  2008-05-16
4974         NaN  2006-04-30
4975         NaN  2016-01-16
4976  2008-03-05  2008-03-08
4977         NaN  2010-03-15
4978  2016-02-10  2016-02-12
4979  2015-06-17  2015-06-19
4980         NaN  2012-04-14
4981         NaN  2013-04-20
4982         NaN         NaN
4983         NaN  2013-04-20
4984         NaN  2009-02-28
4985         NaN  2006-09-16
4986  2015-04-09  2015-04-10
4987         NaN  2005-04-02
4988         NaN  2016-02-12
4989  2015-03-31  2015-04-04
4990         NaN  2006-04-22
4991         NaN  2015-08-17
4992         NaN  2006-07-15
4993         NaN  2006-07-22
4994  2015-04-20  2015-04-24
4995         NaN  2008-03-10
4996  2015-04-14  2015-04-20
4997         NaN  2008-05-17
4998         NaN  2006-05-06
4999  2015-11-25  2015-11-27

[5000 rows x 2 columns]

We use `dropna()` to eliminate rows that have NaN or missing information, as these rows cannot provide comparisons.

In [4]:
presale = presale.dropna()
presale

presale_dt   onsale_dt
4     2005-03-02  2005-03-04
11    2015-12-09  2015-12-11
15    2008-04-03  2008-04-05
16    2015-03-19  2015-03-20
17    2008-05-14  2008-05-17
26    2015-02-04  2015-02-06
29    2016-01-12  2016-01-16
35    2016-02-18  2016-02-19
40    2016-02-09  2016-02-12
45    2015-08-27  2015-08-28
48    2008-03-05  2008-03-08
50    2015-05-08  2015-05-09
51    2015-10-14  2015-10-16
53    2015-11-06  2015-11-14
55    2016-02-23  2016-02-26
58    2016-02-10  2016-02-16
60    2016-02-09  2016-02-16
62    2015-03-04  2015-03-06
65    2015-12-04  2016-01-29
70    2015-03-04  2015-03-06
73    2010-05-11  2010-05-14
80    2015-04-01  2015-02-13
81    2015-11-30  2015-12-03
88    2016-02-15  2016-02-22
90    2016-02-23  2016-02-27
92    2015-10-08  2015-10-09
94    2008-02-18  2008-02-23
97    2016-02-17  2016-02-19
100   2016-02-16  2016-02-19
101   2015-10-06  2015-10-09
...          ...         ...
4927  2015-05-12  2015-05-15
4928  2016-01-19  2016-01-22
4929  2016-01-12  2016-01-16
4932  2016-03-01  2016-03-05
4933  2015-09-14  2015-09-19
4934  2015-10-13  2015-10-16
4936  2015-04-23  2015-04-24
4937  2015-03-31  2015-04-04
4938  2008-03-05  2008-03-08
4940  2016-02-17  2016-02-19
4941  2015-04-09  2015-04-10
4947  2016-01-25  2016-01-29
4950  2016-03-02  2016-03-04
4957  2015-12-16  2015-12-17
4959  2016-02-24  2016-02-29
4960  2015-12-09  2015-12-11
4962  2006-02-23  2006-03-04
4963  2008-03-05  2008-03-08
4964  2016-02-03  2016-02-05
4966  2015-06-09  2015-06-12
4968  2006-05-18  2006-05-19
4973  2008-05-14  2008-05-16
4976  2008-03-05  2008-03-08
4978  2016-02-10  2016-02-12
4979  2015-06-17  2015-06-19
4986  2015-04-09  2015-04-10
4989  2015-03-31  2015-04-04
4994  2015-04-20  2015-04-24
4996  2015-04-14  2015-04-20
4999  2015-11-25  2015-11-27

[2103 rows x 2 columns]

We use `dtypes` to determine whether the columns contain strings or datetime objects.

In [5]:
presale.dtypes

presale_dt    object
onsale_dt     object
dtype: object

Both columns use strings, so we use `pd.to_datetime()` to convert the dates into datetimes, and make sure we use `errors='coerce'` so that if errors are encountered, values are returned as NaT rather than stopping the processing.

In [6]:
presale = presale.apply(pd.to_datetime, errors='coerce')
presale.dtypes

presale_dt    datetime64[ns]
onsale_dt     datetime64[ns]
dtype: object

Now we add a column that calculates the numbers of days bewtween presale and onsale dates.

In [7]:
presale['Days Ahead'] = presale['onsale_dt'] - presale['presale_dt']
presale

presale_dt  onsale_dt Days Ahead
4    2005-03-02 2005-03-04     2 days
11   2015-12-09 2015-12-11     2 days
15   2008-04-03 2008-04-05     2 days
16   2015-03-19 2015-03-20     1 days
17   2008-05-14 2008-05-17     3 days
26   2015-02-04 2015-02-06     2 days
29   2016-01-12 2016-01-16     4 days
35   2016-02-18 2016-02-19     1 days
40   2016-02-09 2016-02-12     3 days
45   2015-08-27 2015-08-28     1 days
48   2008-03-05 2008-03-08     3 days
50   2015-05-08 2015-05-09     1 days
51   2015-10-14 2015-10-16     2 days
53   2015-11-06 2015-11-14     8 days
55   2016-02-23 2016-02-26     3 days
58   2016-02-10 2016-02-16     6 days
60   2016-02-09 2016-02-16     7 days
62   2015-03-04 2015-03-06     2 days
65   2015-12-04 2016-01-29    56 days
70   2015-03-04 2015-03-06     2 days
73   2010-05-11 2010-05-14     3 days
80   2015-04-01 2015-02-13   -47 days
81   2015-11-30 2015-12-03     3 days
88   2016-02-15 2016-02-22     7 days
90   2016-02-23 2016-02-27     4 days
92   2015-10-08 2015-10-09     1 days
94   2008-02-18 2008-02-23     5 days
97   2016-02-17 2016-02-19     2 days
100  2016-02-16 2016-02-19     3 days
101  2015-10-06 2015-10-09     3 days
...         ...        ...        ...
4927 2015-05-12 2015-05-15     3 days
4928 2016-01-19 2016-01-22     3 days
4929 2016-01-12 2016-01-16     4 days
4932 2016-03-01 2016-03-05     4 days
4933 2015-09-14 2015-09-19     5 days
4934 2015-10-13 2015-10-16     3 days
4936 2015-04-23 2015-04-24     1 days
4937 2015-03-31 2015-04-04     4 days
4938 2008-03-05 2008-03-08     3 days
4940 2016-02-17 2016-02-19     2 days
4941 2015-04-09 2015-04-10     1 days
4947 2016-01-25 2016-01-29     4 days
4950 2016-03-02 2016-03-04     2 days
4957 2015-12-16 2015-12-17     1 days
4959 2016-02-24 2016-02-29     5 days
4960 2015-12-09 2015-12-11     2 days
4962 2006-02-23 2006-03-04     9 days
4963 2008-03-05 2008-03-08     3 days
4964 2016-02-03 2016-02-05     2 days
4966 2015-06-09 2015-06-12     3 days
4968 2006-05-18 2006-05-19     1 days
4973 2008-05-14 2008-05-16     2 days
4976 2008-03-05 2008-03-08     3 days
4978 2016-02-10 2016-02-12     2 days
4979 2015-06-17 2015-06-19     2 days
4986 2015-04-09 2015-04-10     1 days
4989 2015-03-31 2015-04-04     4 days
4994 2015-04-20 2015-04-24     4 days
4996 2015-04-14 2015-04-20     6 days
4999 2015-11-25 2015-11-27     2 days

[2103 rows x 3 columns]

We use `value_counts()` to get a summarized view of the days.

In [8]:
presale['Days Ahead'].dt.days.value_counts()

 3       542
 2       522
 1       304
 4       269
 7       128
 5        92
 0        53
 6        49
 9        28
 8        23
-47        9
 30        7
 15        7
 10        5
 11        5
 22        4
-55        4
 12        3
 46        2
-115       2
-7         2
 16        2
 727       2
 367       2
 1829      2
-9         2
 25        2
 71        2
 14        1
-72        1
 739       1
 24        1
-7782      1
-2         1
 56        1
 60        1
-62        1
-64        1
 115       1
 76        1
 78        1
-103       1
-81        1
 63        1
-33        1
-31        1
 23        1
-19        1
 365       1
 1761      1
-7817      1
 3653      1
-7715      1
 214       1
-8386      1
 94        1
-334       1
Name: Days Ahead, dtype: int64

We notice at least one entry is a negative number, which indicates bad data.  We also have absurdly high values, which likely indicates bad idea.  We filter out rows that have negative values as well as rows that contain values over 365, since it seems unlikely that the presale would be any more than a year in advance.

In [9]:
presale = presale[((presale['Days Ahead'].dt.days >= 0) & (presale['Days Ahead'].dt.days <= 365))]
presale

presale_dt  onsale_dt Days Ahead
4    2005-03-02 2005-03-04     2 days
11   2015-12-09 2015-12-11     2 days
15   2008-04-03 2008-04-05     2 days
16   2015-03-19 2015-03-20     1 days
17   2008-05-14 2008-05-17     3 days
26   2015-02-04 2015-02-06     2 days
29   2016-01-12 2016-01-16     4 days
35   2016-02-18 2016-02-19     1 days
40   2016-02-09 2016-02-12     3 days
45   2015-08-27 2015-08-28     1 days
48   2008-03-05 2008-03-08     3 days
50   2015-05-08 2015-05-09     1 days
51   2015-10-14 2015-10-16     2 days
53   2015-11-06 2015-11-14     8 days
55   2016-02-23 2016-02-26     3 days
58   2016-02-10 2016-02-16     6 days
60   2016-02-09 2016-02-16     7 days
62   2015-03-04 2015-03-06     2 days
65   2015-12-04 2016-01-29    56 days
70   2015-03-04 2015-03-06     2 days
73   2010-05-11 2010-05-14     3 days
81   2015-11-30 2015-12-03     3 days
88   2016-02-15 2016-02-22     7 days
90   2016-02-23 2016-02-27     4 days
92   2015-10-08 2015-10-09     1 days
94   2008-02-18 2008-02-23     5 days
97   2016-02-17 2016-02-19     2 days
100  2016-02-16 2016-02-19     3 days
101  2015-10-06 2015-10-09     3 days
104  2015-10-20 2015-10-23     3 days
...         ...        ...        ...
4927 2015-05-12 2015-05-15     3 days
4928 2016-01-19 2016-01-22     3 days
4929 2016-01-12 2016-01-16     4 days
4932 2016-03-01 2016-03-05     4 days
4933 2015-09-14 2015-09-19     5 days
4934 2015-10-13 2015-10-16     3 days
4936 2015-04-23 2015-04-24     1 days
4937 2015-03-31 2015-04-04     4 days
4938 2008-03-05 2008-03-08     3 days
4940 2016-02-17 2016-02-19     2 days
4941 2015-04-09 2015-04-10     1 days
4947 2016-01-25 2016-01-29     4 days
4950 2016-03-02 2016-03-04     2 days
4957 2015-12-16 2015-12-17     1 days
4959 2016-02-24 2016-02-29     5 days
4960 2015-12-09 2015-12-11     2 days
4962 2006-02-23 2006-03-04     9 days
4963 2008-03-05 2008-03-08     3 days
4964 2016-02-03 2016-02-05     2 days
4966 2015-06-09 2015-06-12     3 days
4968 2006-05-18 2006-05-19     1 days
4973 2008-05-14 2008-05-16     2 days
4976 2008-03-05 2008-03-08     3 days
4978 2016-02-10 2016-02-12     2 days
4979 2015-06-17 2015-06-19     2 days
4986 2015-04-09 2015-04-10     1 days
4989 2015-03-31 2015-04-04     4 days
4994 2015-04-20 2015-04-24     4 days
4996 2015-04-14 2015-04-20     6 days
4999 2015-11-25 2015-11-27     2 days

[2061 rows x 3 columns]

We run `value_counts()` again to confirm the filter worked properly.

In [10]:
presale['Days Ahead'].dt.days.value_counts()

3      542
2      522
1      304
4      269
7      128
5       92
0       53
6       49
9       28
8       23
15       7
30       7
11       5
10       5
22       4
12       3
25       2
71       2
16       2
46       2
14       1
365      1
24       1
56       1
60       1
115      1
78       1
94       1
214      1
23       1
63       1
76       1
Name: Days Ahead, dtype: int64

The data looks clean, so we go ahead and use `mean()` to find the average.

In [11]:
presale['Days Ahead'].dt.days.mean()

3.967976710334789

Based on this analysis, we conclude that the average advance notice between regular sales and presales in this dataset is **3.97 days**. 